config/file.go

```diff
@@ -0,0 +1,323 @@
+		return configFile, nil
+	}
+
+	// Otherwise, search for the config/ folder using the same heuristics as above, and build
+	// an absolute path anchored there and joining the given input path (or plain filename).
+	if configFolder, found := fileutils.FindDir("config"); found {
+		return filepath.Join(configFolder, path), nil
+	}
+
+	// Fail altogether if we can't even find the config/ folder. This should only happen if
+	// the executable is relocated away from the supporting files.
+	return "", fmt.Errorf("failed to find config file %s", path)
+}
+
+// Get fetches the current, cached configuration.
+func (fs *fileStore) Get() *model.Config {
+	fs.configLock.RLock()
+	defer fs.configLock.RUnlock()
+
+	return fs.config
+}
+
+// GetEnvironmentOverrides fetches the configuration fields overridden by environment variables.
+func (fs *fileStore) GetEnvironmentOverrides() map[string]interface{} {
+	fs.configLock.RLock()
+	defer fs.configLock.RUnlock()
+
+	return fs.environmentOverrides
+}
+
+// Set replaces the current configuration in its entirety, without updating the backing store.
+func (fs *fileStore) Set(newCfg *model.Config) (*model.Config, error) {
+	fs.configLock.Lock()
+	var unlockOnce sync.Once
+	defer func() {
+		unlockOnce.Do(func() {
+			fs.configLock.Unlock()
+		})
+	}()
+
+	oldCfg := fs.config
+
+	// TODO: disallow attempting to save a directly modified config (comparing pointers). This
+	// wouldn't be an exhaustive check, given the use of pointers throughout the data
+	// structure, but might prevent common mistakes. Requires upstream changes first.
+	// if newCfg == oldCfg {
+	// 	return nil, errors.New("old configuration modified instead of cloning")
+	// }
+
+	newCfg = newCfg.Clone()
+	newCfg.SetDefaults()
+
+	// Sometimes the config is received with "fake" data in sensitive fields. Apply the real
+	// data from the existing config as necessary.
+	desanitize(oldCfg, newCfg)
+
+	if err := newCfg.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "new configuration is invalid")
+	}
+
+	if *oldCfg.ClusterSettings.Enable && *oldCfg.ClusterSettings.ReadOnlyConfig {
+		return nil, ReadOnlyConfigurationError
+	}
+
+	fs.needsSave = true
+
+	// Ideally, Set would persist automatically and abstract this completely away from the
+	// client. Doing so requires a few upstream changes first, so for now an explicit Save()
+	// remains required.
+	// if err := fs.persist(newCfg); err != nil {
+	// 	return nil, errors.Wrap(err, "failed to persist")
+	// }
+
+	fs.config = newCfg
+
+	unlockOnce.Do(func() {
+		fs.configLock.Unlock()
+	})
+
+	// Notify listeners synchronously. Ideally, this would be asynchronous, but existing code
+	// assumes this and there would be increased complexity to avoid racing updates.
+	fs.invokeConfigListeners(oldCfg, newCfg)
+
+	return oldCfg, nil
+}
+
+// Patch merges the given config with the current configuration, without updating the backing store.
```

Comment:
I haven't given this a lot of thought yet, but the rough sketch for how this will work is to accept a `model.Config` on which `SetDefaults` hasn't been called, and then to ignore all `nil` fields.